sdk/python/feast/infra/offline_stores/remote.py

```diff
@@ -0,0 +1,173 @@
+import uuid
+from datetime import datetime
+from pathlib import Path
+from typing import Any, Callable, List, Literal, Optional, Union
+
+import pandas as pd
+import pyarrow as pa
+import pyarrow.flight as fl
+import pyarrow.parquet
+from pydantic import StrictInt, StrictStr
+
+from feast import OnDemandFeatureView
+from feast.data_source import DataSource
+from feast.feature_logging import LoggingConfig, LoggingSource
+from feast.feature_view import FeatureView
+from feast.infra.offline_stores.offline_store import (
+    OfflineStore,
+    RetrievalJob,
+)
+from feast.infra.registry.base_registry import BaseRegistry
+from feast.repo_config import FeastConfigBaseModel, RepoConfig
+from feast.usage import log_exceptions_and_usage
+
+
+class RemoteOfflineStoreConfig(FeastConfigBaseModel):
+    type: Literal["remote"] = "remote"
+    host: StrictStr
+    """ str: remote offline store server port, e.g. the host URL for offline store  of arrow flight server. """
+
+    port: Optional[StrictInt] = None
+    """ str: remote offline store server port."""
+
+
+class RemoteRetrievalJob(RetrievalJob):
+    def __init__(
+        self,
+        config: RepoConfig,
+        feature_refs: List[str],
+        entity_df: Union[pd.DataFrame, str],
+        # TODO add missing parameters from the OfflineStore API
+    ):
+        # Generate unique command identifier
+        self.command = str(uuid.uuid4())
+        # Initialize the client connection
+        self.client = fl.connect(
+            f"grpc://{config.offline_store.host}:{config.offline_store.port}"
+        )
+        # Put API parameters
+        self._put_parameters(feature_refs, entity_df)
+
+    def _put_parameters(self, feature_refs, entity_df):
+        historical_flight_descriptor = fl.FlightDescriptor.for_command(self.command)
+
+        entity_df_table = pa.Table.from_pandas(entity_df)
+        writer, _ = self.client.do_put(
+            historical_flight_descriptor,
+            entity_df_table.schema.with_metadata(
+                {
+                    "command": self.command,
+                    "api": "get_historical_features",
+                    "param": "entity_df",
+                }
+            ),
+        )
+        writer.write_table(entity_df_table)
+        writer.close()
+
+        features_array = pa.array(feature_refs)
```

Comment:
I don't think we should be doing multiple `do_put`s here, the `do_put` should be called for entity dataset only. I would treat features as a parameter and try to pass at along either in the metadata of the previous do_put call or if that proves too hard, I think we will have to encode all the parameters in the FlightDescriptor itself (It could contain uuid and all params, some binary encoding of ours or maybe a proto message). wdyt?

Reply:
I think yes we can pass feature_refs as json string into entity_df_table metadata something like this.        
 ```
       features_json = json.dumps(feature_refs)
        entity_df_table = pa.Table.from_pandas(entity_df)
        writer, _ = self.client.do_put(
            historical_flight_descriptor,
            entity_df_table.schema.with_metadata(
                {
                    "command": self.command,
                    "api": "get_historical_features",
                    "param": "entity_df",
                    "features": features_json
                }
            ),
        )
```

Reply:
Sure, that looks better. I still anticipate some problems with this choice, though. For example, there is an option to pass entity dataset as a query rather than a pandas dataframe (which I'm not sure we'd like to support or not, frankly). In that case, we probably wouldn't do any `do_put` calls whatsover as there's no data to be passed to the server. That's why putting all these into a flight descriptor seems more appropriate to me. Having said that, we can try like this first and revisit the choice later if necessary...

Reply:
We can try to minimize the do_put calls, that's a good catch.
For parameters that have a double nature like `entity_df: Union[pd.DataFrame, str]`, or those that are of primitive type like `project: str, full_feature_names: bool = False`, we can try to encode them in the descriptor as you said, probably together with an identifier of the actual API like `get_historical_features`.
Before trying the proto message option (and the related overhead), I would try to encode them somehow.

Reply:
agreed, proto will be an overkill for now (although probably where we will end up down the line 😄). Just some dict -> json -> base64 or something like that will do.